### PR TITLE
Enable ktlint/detekt tasks on TaskCluster

### PR DESCRIPTION
**Pull request by badboy**
This is a rough draft of bringing back TaskCluster into our setup and enabling just 2 tasks: ktlint and detekt (that way I can hopefully avoid the Rust setup for now).

It's not ready yet, probably needs RelEng help (for setting up the TC side)

### CODE REVIEW

**Comment by badboy** [comment]
@MihaiTabara Would you be able to take a look?
This tries to copy over most of the setup in taskcluster.yml and then only a subset for the decision task and CI tasks. 
In theory it should only run ktlint and detekt (2 gradlew lint tasks).
I rougly replaced all mentions of appservices/app-services/application-services with glean.

Also please advise:
* What do I need to do again to enable TaskCluster on this repository?
* Is there anything you need to do to enable the resources for us?

File: .taskcluster.yml [hunk 1]
```diff
@@ -1,25 +1,21 @@
...
+---
 version: 1
+reporting: checks-v1
 policy:
   pullRequests: public
 tasks:
   $let:
+    taskgraph:
+      branch: taskgraph
+      revision: f949fb56d96b9509b071ba55853f2d60e8dbbf57
```
Comment (MihaiTabara) [inline .taskcluster.yml:13]: This looks good, using an up-to-date revision. I've been on a 4 months old for quite a while :)

File: .taskcluster.yml [hunk 2]
```diff
@@ -1,25 +1,21 @@
...
+    trustDomain: glean
```
Comment (MihaiTabara) [inline .taskcluster.yml:14]: We'll need to ramp-up workers for this. I'll take care of this in https://bugzilla.mozilla.org/show_bug.cgi?id=1628403

File: .taskcluster.yml [hunk 3]
```diff
@@ -36,7 +32,7 @@ tasks:
 
     is_repo_trusted:
       # Pull requests on main repository can't be trusted because anybody can open a PR on it, without a review
-      $if: 'tasks_for in ["github-push", "github-release"] && event.repository.html_url == "https://github.com/mozilla/glean"'
+      $if: 'tasks_for in ["github-push"] && event.repository.html_url == "https://github.com/mozilla/glean"'
```
Comment (MihaiTabara) [inline .taskcluster.yml:35]: Yeah, good point on starting with master pushes to see how this works.

File: .taskcluster.yml [hunk 4]
```diff
@@ -48,168 +44,210 @@ tasks:
...
+                    short_head_branch:
+                      $if: 'head_branch[:11] == "refs/heads/"'
+                      then: {$eval: 'head_branch[11:]'}
+                      else: ${head_branch}
+                  in:
+                    - 'assume:repo:${repoUrl[8:]}:branch:${short_head_branch}'
+                else:
+                  $if: 'tasks_for == "github-pull-request"'
+                  then:
+                    - 'assume:repo:github.com/${event.pull_request.base.repo.full_name}:pull-request'
```
Comment (MihaiTabara) [inline .taskcluster.yml:168]: I'll need to generate all these roles/scopes along the way, before we can make this work. Tracked in the same aforementioned bug.

File: .taskcluster.yml [hunk 5]
```diff
@@ -48,168 +44,210 @@ tasks:
...
+                    - GLEAN_BASE_REPOSITORY: '${baseRepoUrl}'
+                      GLEAN_HEAD_REPOSITORY: '${repoUrl}'
+                      GLEAN_HEAD_REF: '${head_branch}'
+                      GLEAN_HEAD_REV: '${head_sha}'
+                      GLEAN_REPOSITORY_TYPE: git
+                      TASKGRAPH_BASE_REPOSITORY: https://hg.mozilla.org/ci/taskgraph
+                      TASKGRAPH_HEAD_REPOSITORY: https://hg.mozilla.org/ci/${taskgraph.branch}
+                      TASKGRAPH_HEAD_REV: ${taskgraph.revision}
+                      TASKGRAPH_REPOSITORY_TYPE: hg
+                      REPOSITORIES: {$json: {appservices: "Application Services", taskgraph: "Taskgraph"}}
```
Comment (MihaiTabara) [inline .taskcluster.yml:188]: Nit: s/APPservices/GLEAN here? so that we can use `--glean-checkout` below at line 208?

File: .taskcluster.yml [hunk 6]
```diff
@@ -48,168 +44,210 @@ tasks:
...
+                                    platform: gecko-decision
+                              - $if: 'tasks_for in ["github-push", "github-pull-request"]'
+                                then:
+                                    symbol: D
+                    - tasks_for: '${tasks_for}'
           "tasks_for == 'github-push' && head_branch == 'refs/heads/master'":
             $mergeDeep:
               - {$eval: 'default_task_definition'}
+              - scopes:
+                - assume:repo:${event.repository.html_url[8:]}:branch:master
```
Comment (MihaiTabara) [inline .taskcluster.yml:250]: Note to self: to ensure this scope is generated as well.

File: taskcluster/ci/config.yml [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  repositories:
+    glean:
+      name: "Glean"
+  cached-task-prefix: project.glean
+  decision-parameters: 'glean_taskgraph:get_decision_parameters'
+
+workers:
+  aliases:
+    b-linux:
+      provisioner: 'glean-{level}'
```
Comment (MihaiTabara) [inline taskcluster/ci/config.yml:19]: Note to self: I need to generate these as well.

File: taskcluster/ci/config.yml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+      os: linux
+      worker-type: 'b-linux'
+    images:
+      provisioner: 'glean-{level}'
+      implementation: docker-worker
+      os: linux
+      worker-type: 'images'
+
+scriptworker:
+  scope-prefix: project:mozilla:glean:releng
```
Comment (MihaiTabara) [inline taskcluster/ci/config.yml:30]: Note to self: I need to enrich scriptworker to whitelist this in CoT.

**Comment by MihaiTabara** [comment]
> @MihaiTabara Would you be able to take a look?
> This tries to copy over most of the setup in taskcluster.yml and then only a subset for the decision task and CI tasks.
> In theory it should only run ktlint and detekt (2 gradlew lint tasks).
> I rougly replaced all mentions of appservices/app-services/application-services with glean.
> 
> Also please advise:
> 
>     * What do I need to do again to enable TaskCluster on this repository?

You need to enable https://github.com/apps/firefoxci-taskcluster Github app.

>     * Is there anything you need to do to enable the resources for us?

Yeah, I need to add the necessary resources for `glean` under https://hg.mozilla.org/ci/ci-configuration so that we can automatically generate all the scopes/roles/workers and such and apply those against Taskcluster database so that we can start using this.

**Comment by MihaiTabara** [comment]
FWIW - taking care of the necessary TC resources in https://bugzilla.mozilla.org/show_bug.cgi?id=1629817

**Comment by badboy** [comment]
Fixed the one outstanding issue and requested taskcluster support through https://github.com/apps/firefoxci-taskcluster.

I think that needs approval from Mozilla/GitHub admins first.

**Comment by MihaiTabara** [comment]
> Fixed the one outstanding issue and requested taskcluster support through https://github.com/apps/firefoxci-taskcluster.
> 
> I think that needs approval from Mozilla/GitHub admins first.

I poked around to speed this up, will be back with some answers.

**Comment by badboy** [comment]
Looks like we have working ktlint and detekt tasks: https://github.com/mozilla/glean/runs/608800547
They are also listed in the "Checks" list on this PR.

taskcluster is enabled repo-wide, and already comments on other PRs that lack the config.

Right now this _only_ runs 2 tasks: ktlint and detekt.
It runs on all pushes to this repository, including the master branch.
Thanks @MihaiTabara for all the help.

@Dexterp37 I propose to merge this, we can then move forward with additional tasks on TC to eventually get to releases there.

File: taskcluster/docker/linux/Dockerfile [hunk 9]
```diff
@@ -86,65 +99,68 @@ RUN locale-gen en_US.UTF-8
...
         "build-tools;${ANDROID_BUILD_TOOLS}" \
         "extras;android;m2repository" \
         "extras;google;m2repository"
 
+RUN chown -R worker:worker /builds/worker/android-sdk
+
 # Android NDK
 
-ENV ANDROID_NDK_VERSION "r19c"
+ENV ANDROID_NDK_VERSION "r20"
```
Comment (Dexterp37) [inline taskcluster/docker/linux/Dockerfile:125]: Should we document somewhere that this is one location to update if we'll end up bumping the NDK version? (can happen in a separate PR)

File: taskcluster/docker/linux/Dockerfile [hunk 10]
```diff
@@ -86,65 +99,68 @@ RUN locale-gen en_US.UTF-8
...
 RUN curl -sfSL --retry 5 --retry-delay 10 https://dl.google.com/android/repository/android-ndk-${ANDROID_NDK_VERSION}-linux-x86_64.zip > ndk.zip \
-    && unzip -q ndk.zip -d /build \
+    && unzip -q ndk.zip -d /builds/worker \
     && rm ndk.zip \
-    && mv /build/android-ndk-${ANDROID_NDK_VERSION} ${ANDROID_NDK_ROOT}
+    && mv /builds/worker/android-ndk-${ANDROID_NDK_VERSION} ${ANDROID_NDK_ROOT}
 
-ENV ANDROID_NDK_TOOLCHAIN_DIR /root/.android-ndk-r19c-toolchain
+ENV ANDROID_NDK_TOOLCHAIN_DIR /builds/worker/.android-ndk-r20-toolchain
 ENV ANDROID_NDK_API_VERSION 21
```
Comment (Dexterp37) [inline taskcluster/docker/linux/Dockerfile:137]: Can we document this somewhere, as that we easily keep this consistent if we ever update?

File: .buildconfig.yml [hunk 11]
```diff
@@ -14,6 +14,7 @@ projects:
     path: samples/android/app
     artifactId: glean-sample-app
     description: 'An app demoing how to use the Glean library to collect and send telemetry data.'
+    publications: []
```
Comment (Dexterp37) [inline .buildconfig.yml:17]: What is this doing? Making sure we never end up publishing the artifacts? If so, would you kindly add a comment? I was not able to find the reference syntax from this file, I'm assuming this is Mozilla-specific stuff?

File: taskcluster/docker/linux/Dockerfile (hunk 9 above)
Comment (badboy) [reply taskcluster/docker/linux/Dockerfile:125]: We should. I'll take care of that

File: taskcluster/docker/linux/Dockerfile (hunk 10 above)
Comment (badboy) [reply taskcluster/docker/linux/Dockerfile:137]: Yup, same bug as above

File: .buildconfig.yml (hunk 11 above)
Comment (badboy) [reply .buildconfig.yml:17]: The taskcluster decision task now parses this file to find things.
It looks for a `publications` item and explodes if that is missing.
I could have also fixed the code over there to handle it more gracefully but I also wanted to keep it as close as possible to a-s' config, so it's easier to compare.

Comment (Dexterp37) [reply .buildconfig.yml:17]: I see. Consider just adding a note saying that and pointing to any issue filed for taskcluster, if avaialble.

File: taskcluster/docker/linux/Dockerfile (hunk 9 above)
Comment (badboy) [reply taskcluster/docker/linux/Dockerfile:125]: https://bugzilla.mozilla.org/show_bug.cgi?id=1632440

**Comment by badboy** [comment]
> I saw that @MihaiTabara left a few comments around as reminders. Were them all addressed/do they need bugs?

Yup, these were mostly about the required permissions, and the tasks running succesfully means they are set up.
Mihai tracks additional stuff in some bugs as well

**Pull request merged**